src/outputs/parthenon_hdf5.cpp

```diff
@@ -798,34 +794,84 @@ void PHDF5Output::WriteOutputFileImpl(Mesh *pm, ParameterInput *pin, SimTime *tm
   std::unique_ptr<hbool_t[]> sparse_allocated(new hbool_t[num_blocks_local * num_sparse]);
 
   // allocate space for largest size variable
-  const hsize_t varSize = nx3 * nx2 * nx1;
-  int vlen_max = 0;
+  int varSize_max = 0;
   for (auto &vinfo : all_vars_info) {
-    vlen_max = std::max(vlen_max, vinfo.vlen);
+    const int varSize =
+        vinfo.nx6 * vinfo.nx5 * vinfo.nx4 * vinfo.nx3 * vinfo.nx2 * vinfo.nx1;
+    varSize_max = std::max(varSize_max, varSize);
   }
 
   using OutT = typename std::conditional<WRITE_SINGLE_PRECISION, float, Real>::type;
-  std::vector<OutT> tmpData(varSize * vlen_max * num_blocks_local);
+  std::vector<OutT> tmpData(varSize_max * num_blocks_local);
 
   // create persistent spaces
   local_count[0] = num_blocks_local;
   global_count[0] = max_blocks_global;
-  local_count[2] = global_count[2] = nx3;
-  local_count[3] = global_count[3] = nx2;
-  local_count[4] = global_count[4] = nx1;
+  local_count[4] = global_count[4] = nx3;
+  local_count[5] = global_count[5] = nx2;
+  local_count[6] = global_count[6] = nx1;
 
   // for each variable we write
   for (auto &vinfo : all_vars_info) {
     // not really necessary, but doesn't hurt
     memset(tmpData.data(), 0, tmpData.size() * sizeof(OutT));
 
     const std::string var_name = vinfo.label;
-    const hsize_t vlen = vinfo.vlen;
     const hsize_t nx6 = vinfo.nx6;
     const hsize_t nx5 = vinfo.nx5;
     const hsize_t nx4 = vinfo.nx4;
 
-    local_count[1] = global_count[1] = vlen;
+    local_count[1] = global_count[1] = nx6;
+    local_count[2] = global_count[2] = nx5;
+    local_count[3] = global_count[3] = nx4;
+
+    std::vector<hsize_t> alldims({nx6, nx5, nx4, static_cast<hsize_t>(vinfo.nx3),
+                                  static_cast<hsize_t>(vinfo.nx2),
+                                  static_cast<hsize_t>(vinfo.nx1)});
+
+    int ndim = -1;
+    if (vinfo.where == MetadataFlag(Metadata::Cell)) {
+      ndim = 3 + vinfo.ndim + 1;
+      for (int i = 0; i < vinfo.ndim; i++) {
+        local_count[1 + i] = global_count[1 + i] = alldims[3 - vinfo.ndim + i];
+      }
+      local_count[vinfo.ndim + 1] = global_count[vinfo.ndim + 1] = nx3;
+      local_count[vinfo.ndim + 2] = global_count[vinfo.ndim + 2] = nx2;
+      local_count[vinfo.ndim + 3] = global_count[vinfo.ndim + 3] = nx1;
+
+#ifndef PARTHENON_DISABLE_HDF5_COMPRESSION
+      if (output_params.hdf5_compression_level > 0) {
+        // we need chunks to enable compression
+        const std::array<hsize_t, H5_NDIM> chunk_size(
+            {1, 1, 1, 1, static_cast<hsize_t>(nx3), static_cast<hsize_t>(nx2),
+             static_cast<hsize_t>(nx1)});
+        PARTHENON_HDF5_CHECK(H5Pset_chunk(pl_dcreate, ndim, chunk_size.data()));
```

Comment:
```suggestion
        PARTHENON_HDF5_CHECK(H5Pset_chunk(pl_dcreate, ndim, chunk_size.data()+(7-ndim)));
```
Very quick hack but this worked to fix the failing test in AthenaPK. This would have to be fixed on the line below too

Reply:
Why does this fix work?
This looks to me like there's a discrepancy between been the chuck size definition and the slab definition.
IIRC we always want chunks that cover the data of a variable (component) of the entire block.

Reply:
It looks to me like the fix is either @forrestglines fix or the array size should instead always be 7 in the set chunk call?

Reply:
From what I now understand about this commit, this bug, and why it's occurred here:

With both this commit and previous versions (output version 3 vs output version 2) AthenaPK will always output 5D data -- `block,var,k,j,i`. Other than the version increase from 2 to 3, the data format for AthenaPK does not change with this commit, which I've verified with `h5diff` for one case.

The chunk size here has been incorrectly constructed in the new commit, however. It now constructs the array `{1,1,1,nvar,nz,ny,nx}` for AthenaPK then tells HDF5 to use the first 5 dimensions for chunks, in other words, `{1,1,1,nvar,nz}` for chunks.

In this particular example we use a 32x32x8 grid with 9 variables, leading to a chunk size of `{1,1,1,9,32}`. We then write `{nblocks,9,32,32,8}` of data, which HDF5 apparently dislikes because that last dimension is smaller than the chunk dimension. Maybe this is the only test where `ndim<7` and `nz>nx`, which lead to this chunk error?

My fix is a dirty pointer arithmetic fix to push HDF5 to actually use the last 5 dimensions for a chunk size, or `{1,nvar,nz,ny,nx}`. A better fix would be to construct the array with `ndim` dimensions in the first place.

@Yurlungur  I first tried to increase the chunk dimensions to 7 to use all the array for the chunk size, which also caused HDF5 to fail.

Reply:
I think your fix makes sense.

Reply:
Your analysis makes sense @forrestglines 

Why does the higher dim chuncking not work?
Sounds like a bug in HDF5?

I think in terms of chunks we should force chunks of `{1,.., 1,nz,ny,nx}`  (or more specifically `{other potential dims set to 1} + {nz, ny, ,nx}`) because compression will work over individual chunks.
This way we'd compress always over a single component (scalar, vector or tensor component).

Reply:
I don't think it's a bug, it's reasonable for HDF5 to require the chunk dimension to match or at least not exceed the data dimension.

For our compression purposes in AthenaPK, I'd imagine there's a lot of compression to be made within the variable vector since velocities in different dimensions and magnetic fields in different dimensions should be similar

Reply:
@forrestglines I agree with what you say above, thanks for identifying and solving this bug. I've pushed a fix where we construct `chunk_size` to have the correct entries in the first `ndim` array elements. It is still sized to `H5_NDIM` at compile time. Could I trouble you to re-run the AthenaPK test to see that it still works (or let me know which test failed previously on chunking and I can try)?

I'm not sure it's an HDF5 bug if the always-7-dimensional chunk size was failing, since the actual output datasets are not always 7-dimensional and I assume chunk and dataset need to have the same dimension. I could be wrong though.

Reply:
The test suite just passed - this does fix the issue.

Reply:
Thanks @forrestglines